Platform normalization

Ericson2314 opened the pull request:
###### Motivation for this change

Finally deal with the awkward friction between `{build,host,target}Platform.system` and `{build,host,target}Platform.config`. This is a followup to https://github.com/NixOS/nixpkgs/pull/22575 and fixes https://github.com/NixOS/nixpkgs/issues/24170 .

@nbp this uses your old system code and tests. I think I've used the module type system idiomatically, but I am unsure about the "CPU families" bit. In lieu of real sets, I used a map [why do we say attribute set"...], but perhaps `matchAttrs` should be mutually recursive with a `matchLists` that treats the list as a set (or looks for sublists].

~~@nbp Also, I broke the existing module system test infra with some mysterious call package exception. Not sure what's up with that, but want to get this up for review.~~ Fixed.

This can be reviewed or even merged commit-by-commit, but watch out: github is putting the commits out or order again (argh!).

CC @wizeman 

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 committed:
lib: Collect system/platform related files

Previously, platforms was a random thing in top-level

Ericson2314 committed:
lib: Fix system parsing, and use for doubles lists

The old hard-coded lists are now used to test system parsing.

In the process, make an `assertTrue` in release lib for eval tests; also
use it in release-cross

Ericson2314 committed:
top-level: Less indirection for lib in release*.nix

Ericson2314 committed:
release-cross: For fuloong, just use the predefined platform

This is a semantic change, but probably a safe one. In any event, this is
very old hardware that probably no one uses anymore anyways.

Ericson2314 committed:
Elaborate `localSystem` and `crossSystem` in a consistent manner

Ericson2314 committed:
lib: Fix `matchAttrs`

Ericson2314 committed:
lib: Fix preexisting bugs in old platforms code (now bugs in tests)

Warning, this changes the compatibility claims of existing packages!

Mic92 commented on pkgs/top-level/release-lib.nix:
```diff
@@ -46,6 +46,11 @@ rec {
+  assertTrue = bool:
+    if bool
+    then pkgs.runCommand "evaluated-to-true" {} "true"
+    then pkgs.runCommand "evaluated-to-false" {} "false";
```
this looks like an syntax error.

Mic92 replied:
ok this was fixed later.

Ericson2314 replied:
Well, I want each commit to stand on its own so that should be correct from the start!

Ericson2314 replied:
Fixed from the start!

Mic92 commented on lib/tests/systems.nix:
```diff
@@ -14,10 +14,10 @@ let mseteq = x: y: lib.sort lib.lessThan x == lib.sort lib.lessThan y; in
 
-  arm = assertTrue (mseteq arm [ "armv5tel-linux" "armv6l-linux" "armv7l-linux" ]);
-  i686 = assertTrue (mseteq i686 [ "i686-linux" "i686-freebsd" "i686-netbsd" "i686-cygwin" ]);
+  arm = assertTrue (mseteq arm [ "armv5tel-linux" "armv6l-linux" "armv7l-linux" "aarch64-linux" ]);
```
I think there was a commit, which removed explicitly aarch64 from arm, because the cpu architecture is different.

Mic92 commented on lib/tests/systems.nix:
```diff
@@ -0,0 +1,31 @@
+{
+  all = assertTrue (mseteq all (linux ++ darwin ++ cygwin ++ freebsd ++ openbsd ++ netbsd ++ illumos));
+
+  arm = assertTrue (mseteq arm [ "armv5tel-linux" "armv6l-linux" "armv7l-linux" "aarch64-linux" ]);
```
I think there was a commit, which removed explicitly aarch64 from arm, because the cpu architecture is different.

Mic92 replied:
This was a241abf603e7fe971bdddc7c9454581e26c15d5a

Ericson2314 replied:
Just as I include `x86_64` in the `x86` family, so I still include `aarch64` in the `arm` one. But the `arm` list there also constrains the word width to 32 bits---that should do the trick.

shlevy commented:
Just quickly glanced through this, why are we using the module system type system here? This seems like a really complex change and it's not clear to me why we're doing it this way.

Ericson2314 commented:
@shlevy Well, on some level, I used it because the code that already existed by @nbp and wanting to learn more about it :).

But, assuming we indeed do want to have a white list of known components rather than plain strings, I don't actually think this could be simplified much.

Ericson2314 commented:
Ok fixed the module tests (mistake in 2nd commit).

Ericson2314 commented:
@nbp You have time to review this? Besides use of the module system, it would be good to figure out some plan for all the crazy edge cases when converting between LLVM triples and Nix "doubles" (e.g. stuff like `darwin10` for OS, windows in general).

edolstra commented on pkgs/top-level/release-lib.nix:
```diff
@@ -43,6 +46,11 @@ rec {
+  assertTrue = bool:
+    if bool
+    then pkgs.runCommand "evaluated-to-true" {} "touch $out"
+    else pkgs.runCommand "evaluated-to-false" {} "false";
```
Why is this? Why not just use `assert`?

wmertens replied:
@Ericson2314 so what is easier, an extra job to test these, or changing Hydra so it tracks evaluation errors by revision?

Ericson2314 replied:
Hmm? In a later PR I rewrote the tests the semi-recursive-Nix way.

edolstra commented on lib/systems/parse.nix:
```diff
@@ -0,0 +1,167 @@
+                         , abi    ? assert false; null # ditto
+                         } @ args: let
+    getCpu = name:
+      attrByPath [name] (throw "Unknow cpuType `${name}'.")
```
`Unknow` -> `Unknown`. Also don't use backticks as quote characters please.

Ericson2314 replied:
Thanks

nbp replied:
@edolstra the backtick-quote used to be the way names were reported by Nix tools.

edolstra commented on lib/trivial.nix:
```diff
@@ -108,6 +108,9 @@ rec {
   flip = f: a: b: f b a;
 
+  # Apply function if argument is non-null
+  mapNullable = f: a: if isNull a then a else f a;
```
Functions used in only one place shouldn't be added to `lib`.

Ericson2314 replied:
I can do that, but I've wanted this many times. I bet I could grep for `== null|isNull` and probably find many more uses. I'd be happy to do so after this PR to prove its utility :).

Ericson2314 replied:
https://github.com/NixOS/nixpkgs/pull/24974 other uses :D.

edolstra commented:
To be honest I've never understood what `lib/systems.nix` etc. are for. Can anybody explain what the goal is?

If the goal is to provide a handful of properties of platforms, it would be simpler to just enumerate them, given that we only support something like 6 platforms. E.g.
```
platformProperties = {
  "i686-linux" = {
    is64Bit = false;
    execFormat = "elf";
    ...
  };
  "x86_64-darwin" = {
    is64Bit = true;
    execFormat = "macho";
    ...    
  };
}
```
Though some properties seem questionable. For example, why do we care whether a platform is 64 bit or big endian? (What do we even mean with 64 bit? LP64, LLP64, ...?)

The other thing I've never understood is why `pkgs/top-level/platforms.nix` (`nixpkgs/lib/systems/platforms.nix` in this PR) contains Linux kernel configuration. Kernel configuration should really be moved to `pkgs/os-specific/linux/kernel` (maybe just added to `common-config.nix`), and most can be eliminated (e.g. lines like `IPV6 m` are not platform-specific).

nbp commented on lib/systems/parse.nix:
```diff
@@ -26,101 +25,143 @@ rec {
-    l = lib.splitString "-" s;
-
+  mkSystemFromSkeleton = { cpu
+                         , vendor ? assert false; null # optional, but fallback too complex for here
```
nit: These are actually inferred below, update the comment to say so.

Ericson2314 replied:
Done

Ericson2314 commented on pkgs/top-level/release-lib.nix:
```diff
@@ -43,6 +46,11 @@ rec {
+  assertTrue = bool:
+    if bool
+    then pkgs.runCommand "evaluated-to-true" {} "touch $out"
+    else pkgs.runCommand "evaluated-to-false" {} "false";
```
Basically a workaround around hydra not tracking evaluation errors per revisions. I'd like this to count towards the "test failed vs test pass" numbers like everything else.

edolstra replied:
IMHO such tests are better done at build time (i.e. by having a single job that does `nix-instantiate --eval tests.nix`, probably by adding it to `lib/tests/release.nix`). I don't want to clutter Nixpkgs/NixOS jobsets with lots of tiny Nixpkgs lib tests...

nbp replied:
This is basically what the module tests are doing, they basically spawn multiple Nix expressions tests within one job which contains all module tests.

Ericson2314 replied:
I feel like recursive nix is a very heavy handled solution here. If we don't want sub jobs so flattened, hydra should change the way it treats sub jobs.

nbp replied:
Note that this is not what I would call recursive nix, as the module system does not evaluate any derivation, it only use the interpreter with `nix-instanciate` command, and check for the output of the command.

Ericson2314 commented:
> To be honest I've never understood what lib/systems.nix etc. are for. Can anybody explain what the goal is?

My end goal here is to eventually replace all the `stdenv.is*` predicates, which are woefully inadequate/misleading for cross compilation.

> If the goal is to provide a handful of properties of platforms, it would be simpler to just enumerate them, given that we only support something like 6 platforms. E.g.

While we only support a few *build* platforms, people can and do specify all sorts of obscure stuff for the *host* platform. I want to be more general so the same predicates can be used for all of build, host, and target.

> For example, why do we care whether a platform is 64 bit or big endian? (What do we even mean with 64 bit? LP64, LLP64, ...?)

What CPU attributes we track is less important than a mechanism to do so. If this specific needs to be given more detailed or removed, I'd be happy to do so.

> The other thing I've never understood is why pkgs/top-level/platforms.nix (nixpkgs/lib/systems/platforms.nix in this PR) contains Linux kernel configuration.

Yeah I agree that file / the `{build,host.target}Platform.platform` (formerly `crossPlatform.platform`) attribute set is an undocumented, unorganized cesspool. I think this is a step in the right direction---at least get the file in a sane location before we triage its contents. The only kernel config I'd even consider keeping is that which effects the contents of the headers---that way downstream derivations can require that some kernel feature be enabled. Everything else kernel-wise deserves a better home and I like your suggestion.

nbp commented on lib/systems/default.nix:
```diff
@@ -2,4 +2,22 @@ rec {
+  elaborate = args: let
+    final = {
+      # Prefer to parse `config` as it is strictly more informative.
+      parsed = parse.mkSystemFromString (if args ? config then args.config else args.system);
```
What is this `config` attribute and from where does it come from?

wmertens replied:
Sadly that wiki page is gone now - Can we not rename this attribute to `llvmTargetConfig`?

Ericson2314 replied:
https://web.archive.org/web/20160322211219/https://nixos.org/wiki/CrossCompiling

I'm all for improving the name but we must be careful as this is a public interface to how Nixpkgs is instantiated.

Also, while I follow LLVM most closely, other tools use these too.

wmertens replied:
ah right. In that case, maybe rename `args` to `compileConfig`?

Ericson2314 commented on lib/systems/default.nix:
```diff
@@ -2,4 +2,22 @@ rec {
+  elaborate = args: let
+    final = {
+      # Prefer to parse `config` as it is strictly more informative.
+      parsed = parse.mkSystemFromString (if args ? config then args.config else args.system);
```
It's an LLVM target triple. The name is unfortunate but my doing. See the code block in https://nixos.org/wiki/CrossCompiling#Cross-compiling_in_practice for example.

Ericson2314 commented on lib/attrsets.nix:
```diff
@@ -424,11 +424,11 @@ rec {
    */
-  matchAttrs = pattern: attrs:
-    fold or false (attrValues (zipAttrsWithNames (attrNames pattern) (n: values:
+  matchAttrs = pattern: attrs: assert isAttrs pattern;
```
Done

Ericson2314 commented on lib/systems/parse.nix:
```diff
@@ -26,101 +25,143 @@ rec {
-    };
+  isKernel = x: isType "kernel" x
+    && isExecFormat x.execFormat
+    && all isKernelFamily (attrValues x.families);
```
Inductive enforcement makes sense. Likewise for `isCpuType` which you wrote I'd think :D?

nbp replied:
yes, please.

Ericson2314 replied:
Done

Ericson2314 commented on lib/systems/parse.nix:
```diff
@@ -26,101 +25,143 @@ rec {
-    inherit name cpu arch kernel;
+    && isVendor  x.vendor
+    && isKernel  x.kernel
+    && isAbi     x.abi;
```
Done

Ericson2314 commented:
I've addressed all concerns but the nature of the tests. There is some future worked blocked on this I'd like to get to, so I am merging now. @edolstra if you'd like me to redo the tests to match what @nbp did, just ping and I'll make another PR doing that.

cstrahan commented on lib/systems/parse.nix:
```diff
@@ -0,0 +1,173 @@
+                         } @ args: let
+    getCpu = name:
+      attrByPath [name] (throw "Unknown CPU type: ${name}")
+        cpuTypes;
```
@Ericson2314 This might have been pointed out, but... I believe you can also do this to the same effect:

`cpuTypes.${name} or (throw "Unknown CPU type: ${name}")`

At least, if I'm not mistaken. Which is better is a matter of opinion, I suppose (though I guess it's also possible that one performs better than the other).

Ericson2314 replied:
That's true and nicer. And the whole function can be inlined too. Blame @nbp who originally wrote the code a decade ago https://github.com/NixOS/nixpkgs/commit/56ed820f84101fc7b7cc47cbb1f57e616f35ac73 :D:D:D

Ericson2314 replied:
Feel free to make another PR---I've got more to stuff to upstream so not sure whether I have time!

edolstra commented:
@Ericson2314 Yes please, it would be nice to have one Hydra job for Nixpkgs lib tests.
